magda-sleuther-format-layerer/src/format-engine/FormatSnapshotEvaluator.ts

```diff
@@ -0,0 +1,58 @@
+import {
+    Snapshot,
+    Snapshots
+} from "@magda/typescript-common/src/format/MeasureSnapShot";
+import {
+    SnapshotJudge,
+    SnapshotJudgeResult
+} from "@magda/typescript-common/src/format/SnapShotEvaluator";
+import {
+    DownloadMeasure,
+    DownloadExtensionMeasure,
+    DcatFormatMeasure
+} from "./measures";
+import * as _ from "lodash";
+
+//TODO replace hierarchy implementation with a proper Machine Learning Implementation
+//TODO add unit tests
+export class FormatSnapshotJudge extends SnapshotJudge {
+    constructor(snapshots: Snapshots) {
+        super(snapshots);
+    }
+
+    public getBestSnapshot(): SnapshotJudgeResult {
+        // turn array into an object:
+        // TODO change architecture to make this solution better
+        let snapshotDownload: Snapshot;
+        let snapshotExtension: Snapshot;
+        let snapshotDcat: Snapshot;
+
+        this.snapshots.snapshots.forEach(function(snapshot) {
+            if (snapshot.measure.canDeduceFormat) {
+                if (snapshot.measure instanceof DownloadMeasure)
```

Comment:
Instead of using `instanceof` could we let each `Measure` implementation contain its own confidence level and return it from a `confidenceLevel()`function defined on the `Measure` interface?

Reply:
You stated inside the SnapshotJudge getBestSnapshot() function, that I should get the snapshot with the best confidence level, the problem is, the confidenceLevel inside all the snapshots were determined by each measure. The measure cannot determine accurately how confident it is within itself. The reason is hard to explain, but its because each measure is relying on an assumption that the method its using to determine the format is 100% correct, and it changes its confidence level depending on how much it could use its method. An example is the mimeType npm package. The Measure is assuming that, the package will return a 100% correct format, and will only decrease the accuracy if there is an error with the npm package. The Measure isn't self aware, and can figure out how accurate the npm package is, its assuming its 100% correct. Long story short, the confidenceLevels for each snapshot are relative (think of physics class, where you have different frames of references. If you go inside a rocket in space without windows, you cannot tell how fast its moving if it isn't accelerating). The SnapshotJudge's job is to go meta and determine the absolute confidenceLevel of each Measure. So its wrong to use confidenceLevel instead of instanceof to determine which format to use at this current stage of affairs.